common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts

```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
         }
-
         return null;
     }
+
+    public async getToken(): Promise<RegistryAuthenticationToken> {
+        let authType: string;
+        // Will error out with an internal error if the parameter is not found. This error is determined inside of the
+        // tl.getEndpointAuthorizationScheme/tl.getEndpointAuthorizationParameter and cannot be caught here as it is a
+        // custom error.
+        try {
+            authType = tl.getEndpointAuthorizationScheme(this.endpointName, false);
+        } catch (error) {
+        }
+        try {
+            authType = tl.getEndpointAuthorizationParameter(this.endpointName, "scheme", false);
+        } catch (error) {
+        }
+        if (!authType) {
+            authType = "ServicePrincipal";
+        }
+        if (authType == "ManagedServiceIdentity") {
+            // Parameter 1: retryCount - the current retry count of the method to get the ACR token through MSI authentication
+            // Parameter 2: timeToWait - the current time wait of the method to get the ACR token through MSI authentication
+            return await this._getMSIAuthenticationToken(0, 0);
+        } else {
+            return this.getAuthenticationToken();
+        }
+    }
+
+    private static async _getACRToken(AADToken: string, endpointName: string, registryURL: string, retryCount: number, timeToWait: number): Promise<string> {
+        let deferred = Q.defer<string>();
+        let tenantID = tl.getEndpointAuthorizationParameter(endpointName, 'tenantid', true);
+        let webRequest = new webClient.WebRequest();
+        webRequest.method = "POST";
+        const retryLimit = 5
+        webRequest.uri = `https://${registryURL}/oauth2/exchange`;
+        webRequest.body = (
+            `grant_type=access_token&service=${registryURL}&tenant=${tenantID}&access_token=${AADToken}`
+        );
+        webRequest.headers = {
+            "Content-Type": "application/x-www-form-urlencoded"
+        };
+        webClient.sendRequest(webRequest).then(
+            (response: webClient.WebResponse) => {
+                if (response.statusCode === 200) {
+                    deferred.resolve(response.body.refresh_token);
+                }
+                else if (response.statusCode == 429 || response.statusCode == 500) {
+                    if (retryCount < retryLimit) {
+                        let waitedTime = 2000 + timeToWait * 2;
+                        retryCount += 1;
+                        setTimeout(() => {
+                            deferred.resolve(this._getACRToken(AADToken, endpointName, registryURL, retryCount, waitedTime));
+                        }, waitedTime);
+                    }
+                    else {
+                        deferred.reject(tl.loc('CouldNotFetchAccessTokenforACRStatusCode', response.statusCode, response.statusMessage));
```

Comment:
Is this defined in resources file ? I for some reason only seeing 'MSIFetchError' as defined (unless I am looking at wrong resource file)